ai/core/src/unitycatalog/ai/core/executor/local_subprocess.py

```diff
@@ -0,0 +1,154 @@
+_logger = logging.getLogger(__name__)
+
+
+def _extract_function_name(function_source: str) -> str:
+    """
+    Extracts the name of the first function defined in the provided source code
+    using the ast module.
+    """
+    try:
+        module_ast = ast.parse(textwrap.dedent(function_source))
+    except SyntaxError as e:
+        raise ValueError("Invalid Python source code") from e
+
+    for node in module_ast.body:
+        if isinstance(node, ast.FunctionDef):
+            return node.name
+    raise ValueError("No function definition found in the provided source code.")
+
+
+def _generate_runner_script(function_source: str, cpu_time_limit: int, memory_limit: int) -> str:
+    """
+    Generates a Python runner script that sets resource limits and embeds the function's source code
+    directly into the script that will be executed in a subprocess.
+    """
+    function_name = _extract_function_name(function_source)
+    function_source = textwrap.dedent(function_source)
+    script = f"""
+import sys, base64, json, traceback, asyncio, resource, os
+
+{function_source}
+
+def _limit_resources():
+    try:
+        resource.setrlimit(resource.RLIMIT_CPU, ({cpu_time_limit}, {cpu_time_limit}))
+    except Exception as e:
+        sys.stderr.write("Warning: unable to set RLIMIT_CPU: " + str(e) + "\\n")
+    try:
+        mem_bytes = {memory_limit} * {MB_CONVERSION}
+        resource.setrlimit(resource.RLIMIT_AS, (mem_bytes, mem_bytes))
+    except Exception as e:
+        sys.stderr.write("Warning: unable to set RLIMIT_AS: " + str(e) + "\\n")
+
+def main():
+    _limit_resources()
+    try:
+        b64_params = sys.argv[1]
+        params = __import__('cloudpickle').loads(__import__('base64').b64decode(b64_params))
+        result = {function_name}(**params)
+        if asyncio.iscoroutine(result):
+            result = asyncio.run(result)
+        if result is None:
+            result = "{NO_OUTPUT_MESSAGE}"
+        print(__import__('json').dumps({{"success": True, "result": result}}))
+        sys.stdout.flush()
+    except Exception:
+        tb = __import__('traceback').format_exc()
+        print(__import__('json').dumps({{"success": False, "error": tb}}))
+        sys.stdout.flush()
+
+if __name__ == "__main__":
+    main()
+"""
+    return textwrap.dedent(script)
+
+
+_logger = logging.getLogger(__name__)
+
+
+def run_in_sandbox_subprocess(function_source: str, params: dict[str, Any]) -> tuple[bool, Any]:
+    """
+    Executes a Python source in a sandboxed subprocess.
+
+    The parameters are serialized via cloudpickle and passed to a temporary runner
+    script that applies resource limits and executes the function.
+    The runner returns a JSON message indicating success or error.
+
+    Returns:
+        (success, result): If success is True, result is the function's return value (or a default message if None).
+        Otherwise, result contains the error message.
+    """
+    timeout = float(EXECUTOR_TIMEOUT.get())
```

Comment:
Is float conversion needed here?

Reply:
nope! forgot to update that one after supporting non-string types in the env manager